File: cmd/sorting-alogirthems/insertionSort.go
// Package main implements sorting algorithms including insertion sort
package main

import (
	"fmt"
)

// InsertionSort implements the insertion sort algorithm.
// It works by building a sorted section of the array one element at a time.
// For each element, it finds the appropriate position within the sorted section
// and inserts it there, shifting other elements as necessary.
//
// Time Complexity:
// - Best Case: O(n) - When array is already sorted (minimal comparisons needed)
// - Average Case: O(n²)
// - Worst Case: O(n²) - When array is in reverse order (maximum shifts required)
// Space Complexity: O(1) - Sorts in-place with only a single temporary variable
func InsertionSort(arr []int) {
	n := len(arr)
	
	// Start from the second element (index 1)
	// The first element (index 0) is	 considered already sorted
	for i := 1; i < n; i++ {
		// Store the current element that needs to be inserted
		// into the correct position within the sorted portion
		current := arr[i]
		
		// Start comparing with the elements in the sorted section (to the left)
		j := i - 1
		
		// Move elements of arr[0..i-1] that are greater than current
		// to one position ahead of their current position
		// This creates space to insert the current element
		for j >= 0 && arr[j] > current {
			arr[j+1] = arr[j] // Shift element forward
			j--
		}
		
		// After finding the correct position (j+1),
		// insert the current element there
		arr[j+1] = current
	}
	
	// Print the sorted array
	fmt.Println(arr)
}

File: cmd/sorting-alogirthems/insertionSort.js
"use strict";

// Declare int array
let numbers = [
  55, 10, 305, 220, 175, 90, 330, 25, 145, 250, 70, 195, 385, 40, 275, 130, 350,
  115, 5, 365, 180, 290, 135, 50, 215, 395, 160, 80, 225, 300, 120, 20, 245,
  375, 110, 60, 320, 165, 210, 285, 35, 150, 400, 140, 270, 95, 340, 15, 200,
  370, 105, 290, 45, 185, 310, 100, 280, 235, 365, 125, 55, 240, 390, 360, 230,
  85, 260, 325, 375, 190, 345, 255, 330, 170, 205, 315, 75, 300, 365, 150,
];

// Declare function insertionSort
const insertionSort = (arr) => {
  for (let i = 1; i < arr.length; i++) {
    // Start from 1, not 0
    let temp = arr[i]; // Store current element
    let j = i - 1; // Start comparing with previous elements

    while (j >= 0 && arr[j] > temp) {
      // Shift elements if greater than temp
      arr[j + 1] = arr[j];
      j--;
    }

    arr[j + 1] = temp;
  }

  console.log(arr);
};

insertionSort(numbers);
